packages/contractkit/src/wallets/signers/aws-hsm-signer.ts

```diff
@@ -0,0 +1,109 @@
+import { ensureLeading0x, publicKeyToAddress, trimLeading0x } from '@celo/utils/lib/address'
+import { KMS } from 'aws-sdk'
+import BigNumber from 'bignumber.js'
+import { ec as EC } from 'elliptic'
+import * as ethUtil from 'ethereumjs-util'
+import { parseBERSignature } from '../../utils/ber-utils'
+import { bigNumberToBuffer, bufferToBigNumber, isCanonical } from '../../utils/signature-utils'
+import { getHashFromEncoded, RLPEncodedTx } from '../../utils/signing-utils'
+import { Signer } from './signer'
+
+const SigningAlgorithm = 'ECDSA_SHA_256'
+const secp256k1Curve = new EC('secp256k1')
+
+function getRecoveryParam(message: Buffer, expectedAddress: string, r: Buffer, s: Buffer): number {
+  for (let i = 0; i < 2; i++) {
+    const recoveredPublicKeyByteArray = ethUtil.ecrecover(message, i + 27, r, s)
+    const publicKeyBuff = Buffer.from(recoveredPublicKeyByteArray)
+
+    const pub = bigNumberToBuffer(bufferToBigNumber(publicKeyBuff), 64)
+    if (publicKeyToAddress(ensureLeading0x(pub.toString('hex'))) === expectedAddress) {
+      return i
+    }
+  }
+
+  throw new Error('unable to recover public key from signature')
+}
+
+export default class AwsHsmSigner implements Signer {
+  private kms: KMS
+  private keyId: string
+  private address: string
+
+  constructor(kms: KMS, keyId: string, address: string) {
+    this.kms = kms
+    this.keyId = keyId
+    this.address = address
+  }
+
+  private async findCanonicalSignature(buffer: Buffer): Promise<{ S: BigNumber; R: BigNumber }> {
+    let S: BigNumber
+    let R: BigNumber
+
+    let flag = true
+    while (flag) {
+      const { Signature: signature } = await this.kms
+        .sign({
+          KeyId: this.keyId,
+          MessageType: 'DIGEST',
+          Message: buffer,
+          SigningAlgorithm,
+        })
+        .promise()
+      const { r, s } = parseBERSignature(signature as Buffer)
+      R = bufferToBigNumber(r)
+      S = bufferToBigNumber(s)
+
+      const N = bufferToBigNumber(secp256k1Curve.curve.n)
+      if (isCanonical(S, N)) {
+        flag = false
+      }
+    }
+
+    return { S: S!, R: R! }
+  }
+
+  private async sign(buffer: Buffer): Promise<{ v: number; r: Buffer; s: Buffer }> {
```

Comment:
wdyt of creating a new type for `{ v: number; r: Buffer; s: Buffer }`? It seems to be used a lot outside of this PR too

Reply:
+1 There's one here, but it should probably be moved to a common signature lib:
https://github.com/celo-org/celo-monorepo/blob/11814bdb3b44bddb3c19a9ec0f439ad45e86b6bc/packages/contractkit/src/utils/azure-key-vault-client.ts#L213